server/semantics-service/src/main/java/edu/illinois/ncsa/incore/service/semantics/controllers/TypeController.java

```diff
@@ -263,42 +271,55 @@ public Response searchType(
     public Response publishType(
         @Parameter(name = "Type uri (name).") Document type) {
-        Space space = spaceRepository.getSpaceByName(this.username);
-
-        String id = this.typeDAO.postType(type);
-
-        // add id to matching space
-        space.addMember(id);
-        spaceRepository.addSpace(space);
-
-        return Response.ok(id).status(200)
-            .build();
+        try {
+            if (authorizer.isUserAdmin(this.groups)) {
+                Space space = spaceRepository.getSpaceByName(this.username);
+                String id = this.typeDAO.postType(type);
+                // add id to matching space
+                space.addMember(id);
+                spaceRepository.addSpace(space);
 
+                return Response.ok(id).status(200)
+                    .build();
+            } else {
+                throw new IncoreHTTPException(Response.Status.FORBIDDEN, this.username + " is not an admin.");
+            }
+        } catch (IncoreHTTPException e){
+            throw e;
+        } catch (Exception e) {
+            throw new IncoreHTTPException(Response.Status.BAD_REQUEST, "Invalid type JSON. " + e);
+        }
     }
 
     @DELETE
-    @Path("types/{id}")
+    @Path("types/{name}")
     @Produces(MediaType.APPLICATION_JSON)
-    @Operation(summary = "Delete type by id.")
+    @Operation(summary = "Delete type by name.")
     public Response deleteType(
-        @Parameter(name = "Type id.") @PathParam("id") String id) {
-        String deletedId = this.typeDAO.deleteType(id);
-        if (deletedId == null) {
-            throw new IncoreHTTPException(Response.Status.NOT_FOUND, "Could not find type with id " + id);
-        }
-
+        @Parameter(name = "Type name.") @PathParam("name") String name) {
         // TODO: when this service is not restricted to admins anymore, we will have to check if the user has permissions to delete
-        // remove id from spaces
-        List<Space> spaces = spaceRepository.getAllSpaces();
-        for (Space space : spaces) {
-            if (space.hasMember(deletedId)) {
-                space.removeMember(deletedId);
-                spaceRepository.addSpace(space);
+        if (!authorizer.isUserAdmin(this.groups))
+            throw new IncoreHTTPException(Response.Status.FORBIDDEN, this.username + " is not an admin.");
+
+        if (!this.typeDAO.hasType(name))
+            throw new IncoreHTTPException(Response.Status.NOT_FOUND, "Could not find type with name " + name);
+
+        try {
+            String deletedId = this.typeDAO.deleteType(name);
+            // remove id from spaces
+            List<Space> spaces = spaceRepository.getAllSpaces();
+            for (Space space : spaces) {
+                if (space.hasMember(deletedId)) {
+                    space.removeMember(deletedId);
+                    spaceRepository.addSpace(space);
+                }
             }
+            return Response.ok(deletedId).status(200)
```

Comment:
same here 
might be nicer to return the whole object instead of just the id

Reply:
thanks~~

I changed the DELETE reponse to `return Response.ok(deletedId + " has been deleted.").status(200)
                .build();`

Reply:
ohh i meant to also return the whole object as well.
Just imagine the frontend will need to use that information to dynamically update some of the views; if the return is a sentence `64c15ce0020fd631342887c2 has been deleted.` then the frontend would need to parse it to extract the ID

Reply:
Ah I see. I just change the DELETE response to show the whole object!